// @leapfrogtechnology/sync-db/lib/commands/make.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const config_1 = require("../config");
const io_1 = require("../util/io");
const command_1 = require("@oclif/command");
const fileMakerService = require("../service/fileMaker");
class Make extends command_1.Command {
    /**
     * CLI command execution handler.
     *
     * @returns {Promise<void>}
     */
    async run() {
        const { args, flags: parsedFlags } = this.parse(Make);
        const config = await (0, config_1.loadConfig)(parsedFlags.config);
        const list = await this.makeFiles(config, args.name, parsedFlags.type, {
            create: parsedFlags.create,
            objectName: parsedFlags['object-name']
        });
        for (const filename of list) {
            await (0, io_1.printLine)(`Created ${filename}`);
        }
    }
    /**
     * Make files based on the given name and type.
     *
     * @param {Configuration} config
     * @param {string} filename
     * @param {string} type
     * @param {string} objectName
     * @returns {Promise<string[]>}
     */
    async makeFiles(config, filename, type, options) {
        switch (type) {
            case 'migration':
                return fileMakerService.makeMigration(config, filename, options);
            default:
                throw new Error(`Unsupported file type ${type}.`);
        }
    }
}
Make.description = 'Make migration files from the template.';
Make.args = [{ name: 'name', description: 'Object or filename to generate.', required: true }];
Make.flags = {
    'object-name': command_1.flags.string({
        description: 'Name of table/view/routine to migrate.'
    }),
    create: command_1.flags.boolean({
        default: false,
        description: 'Generate create table stub.'
    }),
    config: command_1.flags.string({
        char: 'c',
        description: 'Custom configuration file.'
    }),
    type: command_1.flags.string({
        char: 't',
        helpValue: 'TYPE',
        description: 'Type of file to generate.',
        default: 'migration',
        options: ['migration', 'view', 'procedure', 'function']
    })
};
exports.default = Make;
