// @leapfrogtechnology/sync-db/lib/commands/migrate-list.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("@oclif/command");
const chalk_1 = require("chalk");
const api_1 = require("../api");
const io_1 = require("../util/io");
const __1 = require("..");
class MigrateList extends command_1.Command {
    constructor() {
        super(...arguments);
        /**
         * Success handler.
         */
        this.onSuccess = async (result) => {
            await (0, io_1.printLine)((0, chalk_1.bold)(` ▸ ${result.connectionId}`));
            const [completedList, remainingList] = result.data;
            const ranCount = completedList.length;
            const remainingCount = remainingList.length;
            // Completed migrations.
            for (const item of completedList) {
                const completedMigrationName = typeof item === 'string' || item instanceof String ? item : item === null || item === void 0 ? void 0 : item.name;
                await (0, io_1.printLine)((0, chalk_1.cyan)(`   • ${completedMigrationName}`));
            }
            // Remaining Migrations
            for (const item of remainingList) {
                await (0, io_1.printLine)((0, chalk_1.grey)(`   - ${item}`));
            }
            if (ranCount === 0 && remainingCount === 0) {
                await (0, io_1.printLine)((0, chalk_1.yellow)('   No migrations.'));
            }
            else if (remainingCount > 0) {
                await (0, io_1.printLine)((0, chalk_1.yellow)(`\n   ${remainingList.length} migrations yet to be run.`));
            }
            else if (remainingCount === 0) {
                await (0, io_1.printLine)('\n   All up to date.');
            }
            await (0, io_1.printLine)();
        };
        /**
         * Failure handler.
         */
        this.onFailed = async (result) => {
            (0, io_1.printLine)((0, chalk_1.bold)((0, chalk_1.red)(` ▸ ${result.connectionId} - Failed`)));
            await (0, io_1.printError)(`   ${result.error}\n`);
        };
    }
    /**
     * CLI command execution handler.
     *
     * @returns {Promise<void>}
     */
    async run() {
        const { flags: parsedFlags } = this.parse(MigrateList);
        const config = await (0, __1.loadConfig)(parsedFlags.config);
        const connections = await (0, __1.resolveConnections)(config, parsedFlags['connection-resolver']);
        const results = await (0, api_1.migrateList)(config, connections, Object.assign(Object.assign({}, parsedFlags), { onSuccess: this.onSuccess, onFailed: this.onFailed }));
        const failedCount = results.filter(({ success }) => !success).length;
        if (failedCount === 0) {
            return process.exit(0);
        }
        (0, io_1.printError)(`Error: Failed retrieving list for ${failedCount} connection(s).`);
        process.exit(-1);
    }
}
MigrateList.description = 'List all the migrations.';
MigrateList.flags = {
    only: command_1.flags.string({
        helpValue: 'CONNECTION_ID(s)',
        description: 'Filter provided connection(s). Comma separated ids eg: id1,id2'
    }),
    'connection-resolver': command_1.flags.string({
        helpValue: 'PATH',
        description: 'Path to the connection resolver.'
    }),
    config: command_1.flags.string({
        char: 'c',
        description: 'Custom configuration file.'
    })
};
exports.default = MigrateList;
