// @leapfrogtechnology/sync-db/lib/commands/prune.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const chalk_1 = require("chalk");
const command_1 = require("@oclif/command");
const api_1 = require("../api");
const __1 = require("..");
const io_1 = require("../util/io");
class Prune extends command_1.Command {
    constructor() {
        super(...arguments);
        /**
         * Started event handler.
         */
        this.onStarted = async (result) => {
            await (0, io_1.printLine)((0, chalk_1.bold)(` ▸ ${result.connectionId}`));
            await (0, io_1.printInfo)('   [✓] Pruning - started');
        };
        /**
         * Success handler.
         */
        this.onSuccess = (result) => (0, io_1.printInfo)(`   [✓] Pruning - completed (${result.timeElapsed}s)\n`);
        /**
         * Failure handler.
         */
        this.onFailed = async (result) => {
            await (0, io_1.printLine)((0, chalk_1.red)(`   [✖] Pruning - failed (${result.timeElapsed}s)\n`));
            await (0, io_1.printError)(`   ${result.error}\n`);
        };
    }
    /**
     * CLI command execution handler.
     *
     * @returns {Promise<void>}
     */
    async run() {
        const { flags: parsedFlags } = this.parse(Prune);
        const isDryRun = parsedFlags['dry-run'];
        const config = await (0, __1.loadConfig)(parsedFlags.config);
        const connections = await (0, __1.resolveConnections)(config, parsedFlags['connection-resolver']);
        if (isDryRun)
            await (0, io_1.printLine)((0, chalk_1.magenta)('\n• DRY RUN STARTED\n'));
        const results = await (0, api_1.prune)(config, connections, Object.assign(Object.assign({}, parsedFlags), { onStarted: this.onStarted, onSuccess: this.onSuccess, onFailed: this.onFailed }));
        const failedCount = results.filter(({ success }) => !success).length;
        if (failedCount === 0) {
            if (isDryRun)
                await (0, io_1.printLine)((0, chalk_1.magenta)('• DRY RUN ENDED\n'));
            return process.exit(0);
        }
        (0, io_1.printError)(`Error: Prune failed for ${failedCount} connection(s).`);
        if (isDryRun)
            await (0, io_1.printLine)((0, chalk_1.magenta)('\n• DRY RUN ENDED\n'));
        process.exit(-1);
    }
}
Prune.description = 'Drop all the synchronized db objects except the ones created via migrations.';
Prune.flags = {
    'dry-run': command_1.flags.boolean({ description: 'Dry run prune.', default: false }),
    only: command_1.flags.string({
        helpValue: 'CONNECTION_ID(s)',
        description: 'Filter provided connection(s). Comma separated ids eg: id1,id2'
    }),
    'connection-resolver': command_1.flags.string({
        helpValue: 'PATH',
        description: 'Path to the connection resolver.'
    }),
    config: command_1.flags.string({
        char: 'c',
        description: 'Custom configuration file.'
    })
};
exports.default = Prune;
