// @leapfrogtechnology/sync-db/lib/migration/source-types/SqlMigrationSourceContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const logger_1 = require("../../util/logger");
/**
 * SQL source migration context for KnexMigrationSource.
 */
class SqlMigrationSourceContext {
    /**
     * SqlMigrationContext constructor.
     *
     * @param {SqlMigrationEntry[]} list
     */
    constructor(list) {
        this.list = list;
        this.log = logger_1.log;
        this.connectionId = '';
        this.log('Instantiated SqlMigrationContext.');
    }
    /**
     * Bind connectionId to the context.
     *
     * @param {string} connectionId
     * @returns {MigrationSourceContext} this
     */
    bind(connectionId) {
        this.connectionId = connectionId;
        this.log = (0, logger_1.dbLogger)(connectionId);
        return this;
    }
    /**
     * Get migration keys.
     *
     * @returns {string[]}
     */
    keys() {
        return this.list.map(({ name }) => name);
    }
    /**
     * Get the migration runner.
     *
     * @param {string} key
     * @returns {MigrationRunner}
     */
    get(key) {
        // TODO: Optimize - no need to find from the list for every item you get. Map it internally.
        const entry = this.list.find(({ name }) => name === key);
        if (!entry) {
            this.log(`Migration entry not found ${key}.`);
            throw new Error(`Cannot find the migration entry ${key}`);
        }
        const logMigration = this.log.extend('migration');
        return {
            up: async (db) => {
                if (entry.queries.up) {
                    logMigration(`UP - ${key}`);
                    return db.raw(entry.queries.up.sql);
                }
                return false;
            },
            down: async (db) => {
                if (entry.queries.down) {
                    logMigration(`DOWN - ${key}`);
                    return db.raw(entry.queries.down.sql);
                }
                return false;
            }
        };
    }
}
exports.default = SqlMigrationSourceContext;
