// @leapfrogtechnology/sync-db/lib/migration/KnexMigrationSource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const logger_1 = require("../util/logger");
/**
 * Migration source class for the Knex Migration API.
 * An instance of this class is passed to the knex's migration config as `migrationSource`.
 *
 * Reference: http://knexjs.org/#Migrations-API
 */
class KnexMigrationSource {
    /**
     * KnexMigrationSource constructor.
     *
     * @param {MigrationEntry[]} migrationLists
     */
    constructor(migrationContext) {
        this.log = (0, logger_1.dbLogger)(migrationContext.connectionId);
        this.migrationContext = migrationContext;
    }
    /**
     * Gets a list of migration names.
     *
     * @returns {Promise<string[]>}
     */
    getMigrations() {
        const migrations = this.migrationContext.keys();
        this.log('getMigrations - resolve:\n%O', migrations);
        return Promise.resolve(migrations);
    }
    /**
     * Gets the name of the migration.
     *
     * @param {string} migration
     * @returns {string}
     */
    getMigrationName(migration) {
        return migration;
    }
    /**
     * Get the migration runner.
     *
     * @param {string} name
     * @returns {Promise<MigrationRunner>}
     */
    getMigration(name) {
        const migration = this.migrationContext.get(name);
        this.log(`getMigration - ${name}`);
        this.log(`getMigration - resolve: %o`, migration);
        return Promise.resolve(migration);
    }
}
exports.default = KnexMigrationSource;
