// @leapfrogtechnology/sync-db/lib/constants.js
"use strict";
/* Global Constants */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DROP_ONLY_OBJECT_TERMINATOR = exports.REQUIRED_ENV_KEYS = exports.DEFAULT_CONFIG = exports.INJECTED_CONFIG_TABLE = exports.CONNECTIONS_FILENAME = exports.CONFIG_FILENAME = exports.NS_PER_SEC = void 0;
// General constants
exports.NS_PER_SEC = 1e9;
// SyncDb specific constants
exports.CONFIG_FILENAME = 'sync-db.yml';
exports.CONNECTIONS_FILENAME = 'connections.sync-db.json';
exports.INJECTED_CONFIG_TABLE = '__sync_db_injected_config';
exports.DEFAULT_CONFIG = {
    basePath: 'src',
    connectionResolver: '',
    execution: 'parallel',
    sql: [],
    hooks: {
        pre_sync: [],
        post_sync: []
    },
    injectedConfig: {
        vars: {}
    },
    migration: {
        directory: 'migration',
        tableName: 'knex_migrations',
        sourceType: 'sql'
    }
};
exports.REQUIRED_ENV_KEYS = ['DB_HOST', 'DB_PASSWORD', 'DB_NAME', 'DB_USERNAME', 'DB_PORT', 'DB_CLIENT'];
exports.DROP_ONLY_OBJECT_TERMINATOR = '.drop';
