// @leapfrogtechnology/sync-db/lib/migration/service/migrator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveJavaScriptMigrations = exports.resolveSqlMigrations = exports.getJavaScriptMigrationNames = exports.getSqlMigrationNames = void 0;
const path = require("path");
const fs_1 = require("../../util/fs");
const sqlRunner_1 = require("../../service/sqlRunner");
const FileExtensions_1 = require("../../enum/FileExtensions");
const FILE_PATTERN_JS = /(.+).js$/;
const FILE_PATTERN_TS = /(.+).ts$/;
const FILE_PATTERN_SQL = /(.+)\.(up|down)\.sql$/;
/**
 * Glob the migration directory and retrieve all the migration entries (names)
 * that needs to be run.
 *
 * Note: The ".up.sql" and ".down.sql" part of the migration files are
 * omitted and and a pair of those files are considered a single migration entry.
 *
 * @param {string} migrationPath
 * @returns {Promise<string[]>}
 */
async function getSqlMigrationNames(migrationPath) {
    const files = await (0, fs_1.glob)(migrationPath);
    const migrationSet = new Set();
    files.forEach(filename => {
        const match = filename.match(FILE_PATTERN_SQL);
        if (match && match.length === 3) {
            migrationSet.add(match[1]);
        }
    });
    return Array.from(migrationSet);
}
exports.getSqlMigrationNames = getSqlMigrationNames;
/**
 * Glob the migration directory and retrieve all the migration entries (names)
 * that need to be run.
 *
 * @param {string} migrationPath
 * @param {string} extension
 * @returns {Promise<string[]>}
 */
async function getJavaScriptMigrationNames(migrationPath, extension) {
    const filenames = await (0, fs_1.glob)(migrationPath);
    const migrationSet = new Set();
    const pattern = extension === FileExtensions_1.default.JS ? FILE_PATTERN_JS : FILE_PATTERN_TS;
    filenames.forEach(filename => {
        const match = filename.match(pattern);
        if (match) {
            migrationSet.add(match[1]);
        }
    });
    return Array.from(migrationSet);
}
exports.getJavaScriptMigrationNames = getJavaScriptMigrationNames;
/**
 * Resolve all the migration source for each of the migration entries using the configurations.
 *
 * @param {string} migrationPath
 * @returns {Promise<SqlMigrationEntry[]>}
 */
async function resolveSqlMigrations(migrationPath) {
    const migrationNames = await getSqlMigrationNames(migrationPath);
    const migrationPromises = migrationNames.map(async (name) => {
        const upFilename = `${name}.up.sql`;
        const downFilename = `${name}.down.sql`;
        const upFileExists = await (0, fs_1.exists)(path.join(migrationPath, upFilename));
        const downFileExists = await (0, fs_1.exists)(path.join(migrationPath, downFilename));
        const up = upFileExists ? await (0, sqlRunner_1.resolveFile)(migrationPath, upFilename) : undefined;
        const down = downFileExists ? await (0, sqlRunner_1.resolveFile)(migrationPath, downFilename) : undefined;
        return {
            name,
            queries: { up, down }
        };
    });
    return Promise.all(migrationPromises);
}
exports.resolveSqlMigrations = resolveSqlMigrations;
/**
 * Resolve all the migration source for each of the migration entries using the configurations.
 *
 * @param {string} migrationPath
 * @param {string} extension
 * @returns {Promise<SqlMigrationEntry[]>}
 */
async function resolveJavaScriptMigrations(migrationPath, extension = FileExtensions_1.default.JS) {
    const migrationNames = await getJavaScriptMigrationNames(migrationPath, extension);
    if (extension === FileExtensions_1.default.TS) {
        // Transpile & execute ts files required on the fly
        require('ts-node').register({
            transpileOnly: true
        });
    }
    const migrationPromises = migrationNames.map(async (name) => {
        const filename = `${name}.${extension}`;
        const { up, down } = require(path.resolve(migrationPath, filename));
        return {
            name: filename,
            methods: {
                up,
                down
            }
        };
    });
    return Promise.all(migrationPromises);
}
exports.resolveJavaScriptMigrations = resolveJavaScriptMigrations;
