// @leapfrogtechnology/sync-db/lib/util/ts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTimestampString = exports.getElapsedTime = void 0;
const constants_1 = require("../constants");
/**
 * Calculate elapsed time from the given start time (process.hrtime()).
 *
 * @param {[number, number]} timeStart
 * @param {(false | number)} fixed
 * @returns {(number | string)}
 */
function getElapsedTime(timeStart, fixed = 2) {
    const timeDiff = process.hrtime(timeStart);
    const timeElapsed = Number(timeDiff[0]) + Number(timeDiff[1] / constants_1.NS_PER_SEC);
    if (fixed === false) {
        return timeElapsed;
    }
    return Number(timeElapsed.toFixed(fixed));
}
exports.getElapsedTime = getElapsedTime;
/**
 * Gets a timestamp string for the given date.
 *
 * @param {Date} [date=new Date()]
 * @returns {string}
 */
function getTimestampString(date = new Date()) {
    const dtf = new Intl.DateTimeFormat('en', {
        year: 'numeric',
        month: '2-digit',
        day: '2-digit',
        hour: '2-digit',
        minute: '2-digit',
        second: '2-digit'
    });
    const [{ value: month }, , { value: day }, , { value: year }, , { value: hour }, , { value: minute }, , { value: second }] = dtf.formatToParts(date);
    return `${year}${month}${day}${hour}${minute}${second}`;
}
exports.getTimestampString = getTimestampString;
